tss-esapi/src/tcti_ldr.rs

```diff
@@ -352,39 +379,67 @@ fn validate_from_str_device_config() {
     assert_eq!(config.path, PathBuf::from("/dev/tpm0"));
 }
 
-/// Configuration for an Mssim TCTI context
+/// Configuration for an Mssim or Swtpm TCTI context
 ///
 /// The default configuration will point to `localhost:2321`
+///
+/// Examples:
+///  - Use of a TCP connection: `swtpm:host=localhost,port=2321`
+///  - Use of a unix socket: `swtpm:path=/path/to/sock`
+///    Available on tpm2-tss >= 3.2.
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub struct NetworkTPMConfig {
-    /// Address of the server to connect to
-    ///
-    /// Defaults to `localhost`
-    host: ServerAddress,
-    /// Port used by the server at the address given in `host`
-    ///
-    /// Defaults to `2321`
-    port: u16,
+pub enum TpmSimulatorConfig {
+    /// Connects to tpm over TCP
+    Tcp {
+        /// Address of the server to connect to
+        ///
+        /// Defaults to `localhost`
+        host: ServerAddress,
+        /// Port used by the server at the address given in `host`
+        ///
+        /// Defaults to `2321`
+        port: u16,
+    },
+    #[cfg(has_simulator_unix_socket)]
+    /// Connects to tpm over Unix socket
+    Unix { path: String },
```

Comment:
:thinking: Correct me if I'm wrong but this makes the `Unix` variant disappear when building on systems without unix sockets support.

Since this variant doesn't have any Unix specific types I'm wondering if that doesn't make writing clients to `rust-tss-esapi` harder. That is if a crate that used `tss-esapi` crate would have to do the detection too (that is writing a build script that propagates `#[cfg(has_simulator_unix_socket)]`).

Wouldn't it be better to actually expose the Unix variant always but only when *using* that variant return an error that this is unsupported?

(That said I'm not completely sure if that's a good approach but I think it's worth considering).

Reply:
This is only for the swtpm which is for simulation / tests. And unix socket is probably even more niche.

Beside that, I'm usually not too comfortable with compile-time checks as the library you link to might be changed from one without support to one with and still not get support. But here, this will be deployed on a nixos so ... my opinion doesn't really matter here.

Up to you folks really.

Reply:
(if we were to expose the unix flavor anyway that would trigger a configuration error sent from the C library which you already have to deal with).

Reply:
I totally agree. This ```cfg``` should be removed. The comment should be updated with information that using this on platforms not supporting unix sockets will result in an error when calling ```TctiContext::initialize```. Then in the the ```initialize function we should check if unix sockets are supported otherwise return an InvalidParam error.

Reply:
Tried by best to document the error case, let me know.

Reply:
I think that is fine. 

I realize that I did not explain my self properly and I apologise for that. So let me explain more thoroughly how 
I meant. What I meant was that the configuration item that you had added would remain but but the check would be moved to ```TctiContyext::initialize``` and the error that would occur would be documented there.


Addition to the ```TctiContyext::initialize``` documentation:
```rust
/// * Errors
///  - Returns InvalidParam error if TpmSimulatorConfig specifies a unix socket and the tpm2-tss
///    version of the library does not support it.
```

The in the initialize function one would add (not sure if you can negate the 'has_unix_socket_support' in cfg_if using 'not' otherwise it should probably be changed so we check if unix socket support is missing):
```rust
cfg_if::cfg_if!{
    if #[cfg(has_unix_socket_support)] {
    } else {
        if let TctiNameConf::Mssim(TpmSimulatorConfig::Unix(_)) = name_conf {
            error!("tpm2-tss library does not support unix sockets.");
            return Err(Error::local_error(WrapperErrorKind::InvalidParam));
        }

        if let TctiNameConf::Swtpm(TpmSimulatorConfig::Unix(_)) = name_conf {
            error!("tpm2-tss library does not support unix sockets.");
            return Err(Error::local_error(WrapperErrorKind::InvalidParam));
        }
    }
}
```

And then the documentation for the enum item could then be something like:
```rust
    /// Connects to TPM over Unix socket
    ///
    /// If used with a library without unix socket support (< 3.2.0),
    /// an InvalidParam error will be returned upon TCTI initialization.
    Unix { path: String },
```

But as I said this is just an example of how I envisioned it and I am fine with merging it if you prefer it your way.

Reply:
There problem with the `#[cfg(has_unix_socket_support)]` is that it looks up the version of `tpm2-tss` at buildtime of that crate. If further down the line we were to upgrade `tpm2-tss` to `>=3.2` it would still return `InvalidParam` and I'm not sure it's the best thing.

Reply:
Yes, I think you are right. The only other option would be to add some kind of tpm2-tss version runtime check to the initialize function or make some kind of analysis of the error returned from ```Tss2_TctiLdr_Initialize``` to see if was caused by using unix sockets with a library that did not support it. But I think that is a thing we could do in another PR if we feel there is a need for it to improve the user experience.